crates/proto/src/xfer/dns_exchange.rs

```diff
@@ -161,52 +250,82 @@ where
-pub struct DnsExchangeConnect<F, S, R>(DnsExchangeConnectInner<F, S, R>)
+/// A wrapper for a future DnsExchange connection.
+///
+/// DnsExchangeConnect is clonable, making it possible to share this if the connection
+///  will be shared across threads.
+pub struct DnsExchangeConnect<F, S, R>(Arc<Mutex<DnsExchangeConnectInner<F, S, R>>>)
 where
     F: Future<Output = Result<S, ProtoError>> + 'static + Send + Unpin,
-    S: DnsRequestSender<DnsResponseFuture = R>,
+    S: DnsRequestSender<DnsResponseFuture = R> + 'static,
     R: Future<Output = Result<DnsResponse, ProtoError>> + 'static + Send + Unpin;
 
 impl<F, S, R> DnsExchangeConnect<F, S, R>
 where
     F: Future<Output = Result<S, ProtoError>> + 'static + Send + Unpin,
-    S: DnsRequestSender<DnsResponseFuture = R>,
+    S: DnsRequestSender<DnsResponseFuture = R> + 'static,
     R: Future<Output = Result<DnsResponse, ProtoError>> + 'static + Send + Unpin,
 {
     fn connect(
         connect_future: F,
         outbound_messages: UnboundedReceiver<OneshotDnsRequest<R>>,
+        sender: DnsRequestStreamHandle<R>,
     ) -> Self {
-        DnsExchangeConnect(DnsExchangeConnectInner::Connecting {
+        DnsExchangeConnect(Arc::new(Mutex::new(DnsExchangeConnectInner::Connecting {
             connect_future,
             outbound_messages: Some(outbound_messages),
-        })
+            sender: Some(sender),
+        })))
+    }
+}
+
+impl<F, S, R> Clone for DnsExchangeConnect<F, S, R>
+where
+    F: Future<Output = Result<S, ProtoError>> + 'static + Send + Unpin,
+    S: DnsRequestSender<DnsResponseFuture = R> + 'static,
+    R: Future<Output = Result<DnsResponse, ProtoError>> + 'static + Send + Unpin,
+{
+    fn clone(&self) -> Self {
+        DnsExchangeConnect(self.0.clone())
     }
 }
 
 impl<F, S, R> Future for DnsExchangeConnect<F, S, R>
 where
     F: Future<Output = Result<S, ProtoError>> + 'static + Send + Unpin,
-    S: DnsRequestSender<DnsResponseFuture = R>,
+    S: DnsRequestSender<DnsResponseFuture = R> + 'static + Send + Unpin,
     R: Future<Output = Result<DnsResponse, ProtoError>> + 'static + Send + Unpin,
 {
     type Output = Result<DnsExchange<S, R>, ProtoError>;
 
-    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
-        self.0.poll_unpin(cx)
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        // TODO: This should really be using the lock() primitive. The issue is the future from that captures
+        //  a lifetime associated with this object, which causes general issues when in managing that future.
+        //  We should come back and evaluate how to migrate to `async fn` for this case.
+        match self.0.try_lock() {
+            Some(mut future) => future.poll_unpin(cx),
+            None => {
+                // TODO: this shouldn't be a hot loop, instead when generators land, use that
+                cx.waker().wake_by_ref();
+                Poll::Pending
```

Comment:
I'm a little concerned about this. At the very least, you might want to consider using an async mutex, like `tokio::sync::Mutex`.

Reply:
this is a futures::lock::Mutex.

The issue here is a lifetime problem. I can't easily capture the value... but I guess we could instead do this:

`self.0.lock().poll()` and then would the task still be properly woken up in that case? Agree that I don't like this.

Reply:
> this is a `futures:lock:Mutex`.

oh, i see, missed that in the diff; my bad.

could this future hold the future returned by `lock()` and poll that until it's ready, or am i missing something?

Reply:
So I ran into a lifetime issue trying to take the `lock()` future and then poll that internally.

In this model the DnsExchangeConnect is cloneable such that it can be handed out to any tasks and they all get a reference to the same DnsExchange once connected. That's why this mutex exists in the Connect object, but is distinct from the one needed for the old background. I think this is what implicitly causes the lifetime issue I was facing, and couldn't think of a way to escape it. But maybe there is a way.

That being said, it's probably not valuable to invest if fixing that if we just go back to an explicit Background.